compiler_gym/datasets/datasets.py

```diff
@@ -0,0 +1,305 @@
+
+        :raises LookupError: If :code:`dataset` is not found.
+        """
+        if dataset is None:
+            if not self._visible_datasets:
+                raise ValueError("No datasets")
+
+            return self._datasets[self.random.choice(list(self._visible_datasets))]
+
+        if isinstance(dataset, Dataset):
+            dataset_name = dataset.name
+        else:
+            dataset_name = resolve_uri_protocol(dataset)
+
+        if dataset_name not in self._datasets:
+            raise LookupError(f"Dataset not found: {dataset_name}")
+
+        return self._datasets[dataset_name]
+
+    def __getitem__(self, dataset: Union[str, Dataset]) -> Dataset:
+        """Lookup a dataset.
+
+        :param dataset: A dataset name, a :class:`Dataset` instance, or
+            :code:`None` to select a dataset randomly.
+
+        :return: A :meth:`Dataset <compiler_gym.datasets.Dataset>` instance.
+
+        :raises LookupError: If :code:`dataset` is not found.
+        """
+        return self.dataset(dataset)
+
+    def __setitem__(self, key: str, dataset: Dataset):
+        self._datasets[key] = dataset
+        if not dataset.hidden:
+            self._visible_datasets.add(dataset.name)
+
+    def __delitem__(self, dataset: Union[str, Dataset]):
+        """Remove a dataset from the collection.
+
+        This does not affect any underlying storage used by dataset. See
+        :meth:`uninstall() <compiler_gym.datasets.Datasets.uninstall>` to clean
+        up.
+
+        :param dataset: A :meth:`Dataset <compiler_gym.datasets.Dataset>`
+            instance, or the name of a dataset.
+
+        :return: :code:`True` if the dataset was removed, :code:`False` if it
+            was already removed.
+        """
+        dataset_name: str = self.dataset(dataset).name
+        if dataset_name in self._visible_datasets:
+            self._visible_datasets.remove(dataset_name)
+        del self._datasets[dataset_name]
+
+    def __contains__(self, dataset: Union[str, Dataset]) -> bool:
+        """Returns whether the dataset is contained."""
+        try:
+            self.dataset(dataset)
+            return True
+        except LookupError:
+            return False
+
+    def benchmarks(self) -> Iterable[Benchmark]:
+        """Enumerate the (possibly infinite) benchmarks lazily.
+
+        Benchmarks order is consistent across runs. One benchmark from each
+        dataset is returned in round robin order until all datasets have been
+        fully enumerated. The order of :meth:`benchmarks()
+        <compiler_gym.datasets.Datasets.benchmarks>` and :meth:`benchmark_uris()
+        <compiler_gym.datasets.Datasets.benchmark_uris>` is the same.
+
+        :return: An iterable sequence of :class:`Benchmark
+            <compiler_gym.datasets.Benchmark>` instances.
+        """
+        return round_robin_iterables((d.benchmarks() for d in self.datasets()))
+
+    def benchmark_uris(self) -> Iterable[str]:
```

Comment:
Should this also have an include_deprecated flag?

Reply:
And, then does that suggest that each dataset probably needs it, too?

Reply:
> And, then does that suggest that each dataset probably needs it, too?

Only entire datasets are deprecated or not, so I don't think we need to propagate the `with_deprecated` to them.